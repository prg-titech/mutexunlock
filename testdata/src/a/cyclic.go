package main

import (
	"sync"
)

func LockInFor(l []int) int {
	var mu1 sync.Mutex

	// LockはじまりでUnlockされないまま自分自身に戻ってきてしまうようなパスがあると検知したい
	// 自分自身もexit nodeと同じようなもの exitnode || 自分自身
	// Lockしているnodeを全部集めてきて，そこからグラフをもう一度辿る
	// Lockをみつるまでのvisitedと，Lockを見つけてからのvisited
	for k, v := range l {
		mu1.Lock()
		if v == 0 {
			// Here it is ok no unlock since mu1 is local, but to alleviate it is better to unlock
			return k // want "missing unlock"
		}
		// want "missng unlock"
	}

	mu1.Unlock()
	return -1 // OK
}
