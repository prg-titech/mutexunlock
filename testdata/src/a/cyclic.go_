package main

import (
	"sync"
)

func _(l []int) int {
	var mu1 sync.Mutex

	for k, v := range l {
		mu1.Lock()
		if v == 0 {
			return k // want "missing unlock"
		} // want "missng unlock"
	}

	mu1.Unlock()
	return -1 // OK
}
