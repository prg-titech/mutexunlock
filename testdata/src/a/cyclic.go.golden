package main

import (
	"fmt"
	"sync"
)

func LockInFor(l []int) int {
	var mu1 sync.Mutex

	for k, v := range l {
		mu1.Lock()
		if v == 0 {
			// Here it is ok no unlock since mu1 is local, but to alleviate it is better to unlock
			mu1.Unlock()
			return k // want "missing unlock"
		}
		mu1.Unlock()
		// want "missng unlock"
	}

	return -1 // OK
}
